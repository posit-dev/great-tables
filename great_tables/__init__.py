# Set version ----
from importlib_metadata import version as _v

__version__ = _v("great_tables")

del _v

# Main gt imports ----

from .gt import GT
from . import vals, loc, style
from ._styles import FromColumn as from_column
from ._helpers import (
    letters,
    LETTERS,
    px,
    pct,
    md,
    html,
    google_font,
    random_id,
    system_fonts,
    define_units,
    nanoplot_options,
)


__all__ = (
    "GT",
    "exibble",
    "letters",
    "LETTERS",
    "px",
    "pct",
    "md",
    "html",
    "google_font",
    "system_fonts",
    "define_units",
    "nanoplot_options",
    "random_id",
    "from_column",
    "vals",
    "loc",
    "style",
)


def __getattr__(k: str):
    # exibble dataset available on top-level module, but is a pandas DataFrame.
    # Since pandas is an optional dependency, we import exibble dynamically.
    if k == "exibble":
        from great_tables.data import exibble

        return exibble

    # allow the data submodule to be accessed directly, as if it were a top-level import
    if k == "data":
        import great_tables.data

        return great_tables.data

    raise AttributeError(f"cannot get attribute {k} from great_tables ({__file__})")
