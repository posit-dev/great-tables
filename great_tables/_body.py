from __future__ import annotations

from typing import TYPE_CHECKING

from ._tbl_data import copy_data

if TYPE_CHECKING:
    from ._gt_data import Body


def body_reassemble(body: Body) -> Body:
    # Note that this used to order the body based on groupings, but now that occurs in the
    # renderer itself.
    return body.__class__(copy_data(body.body))
