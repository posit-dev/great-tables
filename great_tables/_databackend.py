from __future__ import annotations

import importlib
from abc import ABCMeta
from types import ModuleType


def _load_class(mod_name: str, cls_name: str) -> type:
    # follow the path of mods if it is one
    listified_mod_path: list[str] = cls_name.split(".")

    sub_module_path: list[str] = [mod_name] + listified_mod_path[:-1]

    cur_mod: ModuleType = importlib.import_module(sub_module_path[0])
    _cls_name: str = listified_mod_path[-1]

    if len(sub_module_path) == 1:  # no submodules, just the module itself
        return getattr(cur_mod, _cls_name)

    # now, we must increment the module paths to import the final class
    remaining_mods: list[str] = sub_module_path[1:]
    for mod in remaining_mods:
        cur_mod: ModuleType = getattr(cur_mod, mod)

    return getattr(cur_mod, _cls_name)


class _AbstractBackendMeta(ABCMeta):
    def register_backend(cls, mod_name: str, cls_name: str):
        cls._backends.append((mod_name, cls_name))
        cls._abc_caches_clear()


class AbstractBackend(metaclass=_AbstractBackendMeta):
    @classmethod
    def __init_subclass__(cls):
        if not hasattr(cls, "_backends"):
            cls._backends = []

    @classmethod
    def __subclasshook__(cls, subclass):
        for mod_name, cls_name in cls._backends:
            try:
                parent_candidate = _load_class(mod_name, cls_name)
                if issubclass(subclass, parent_candidate):
                    return True
            except (ModuleNotFoundError, ImportError):
                continue

        return NotImplemented
