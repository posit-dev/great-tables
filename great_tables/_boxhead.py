from __future__ import annotations

from typing import TYPE_CHECKING

from ._locations import resolve_cols_c
from ._utils import _assert_list_is_subset
from ._tbl_data import SelectExpr
from ._text import BaseText

if TYPE_CHECKING:
    from ._types import GTSelf


def cols_label(
    self: GTSelf, cases: dict[str, str | BaseText] | None = None, **kwargs: str | BaseText
) -> GTSelf:
    """
    Relabel one or more columns.

    There are three important pieces to labelling:

    * Each argument has the form: {name in data} = {new label}.
    * Multiple columns may be given the same label.
    * Labels may use curly braces to apply special formatting, called unit notation.
      For example, "area ({{ft^2}})" would appear as "area (ftÂ²)".

    See [`define_units()`](`great_tables.define_units`) for details on unit notation.

    Parameters
    ----------
    cases
        A dictionary where the keys are column names and the values are the labels. Labels may use
        [`md()`](`great_tables.md`) or [`html()`](`great_tables.html`) helpers for formatting.

    **kwargs
        Keyword arguments to specify column labels. Each keyword corresponds to a column name, with
        its value indicating the new label.

    Returns
    -------
    GT
        The GT object is returned. This is the same object that the method is called on so that we
        can facilitate method chaining.

    Notes
    -----
    GT always selects columns using their name in the underlying data. This means that a column's
    label is purely for final presentation.

    Examples
    --------

    The example below relabels columns from the `countrypops` data to start with uppercase.

    ```{python}
    from great_tables import GT
    from great_tables.data import countrypops

    countrypops_mini = countrypops.loc[countrypops["country_name"] == "Uganda"][
        ["country_name", "year", "population"]
    ].tail(5)

    (
        GT(countrypops_mini)
        .cols_label(
            country_name="Country Name",
            year="Year",
            population="Population"
        )
    )
    ```

    Note that we supplied the name of the column as the key, and the new label as the value.

    We can also use Markdown formatting for the column labels. In this example, we'll use
    `md("*Population*")` to make the label italicized.

    ```{python}
    from great_tables import GT, md
    from great_tables.data import countrypops

    (
        GT(countrypops_mini)
        .cols_label(
            country_name="Name",
            year="Year",
            population=md("*Population*")
        )
    )
    ```

    We can also use unit notation to format the column labels. In this example, we'll use
    `{{cm^3 molecules^-1 s^-1}}` for part of the label for the `OH_k298` column.

    ```{python}
    from great_tables import GT
    from great_tables.data import reactions
    import polars as pl

    reactions_mini = (
        pl.from_pandas(reactions)
        .filter(pl.col("cmpd_type") == "mercaptan")
        .select(["cmpd_name", "OH_k298"])
    )

    (
        GT(reactions_mini)
        .fmt_scientific("OH_k298")
        .sub_missing()
        .cols_label(
            cmpd_name="Compound Name",
            OH_k298="OH, {{cm^3 molecules^-1 s^-1}}",
        )
    )
    ```
    """
    from great_tables._helpers import UnitStr

    cases = cases if cases is not None else {}
    new_cases = cases | kwargs

    # If nothing is provided, return `data` unchanged
    if len(new_cases) == 0:
        return self

    # Get the full list of column names for the data
    column_names = self._boxhead._get_columns()
    mod_columns = list(new_cases.keys())

    # Stop function if any of the column names specified are not in `cols_labels`
    # msg: "All column names provided must exist in the input `.data` table."
    _assert_list_is_subset(mod_columns, set_list=column_names)

    # Handle units syntax in labels (e.g., "Density ({{ppl / mi^2}})")
    new_kwargs: dict[str, UnitStr | str | BaseText] = {}

    for k, v in new_cases.items():
        if isinstance(v, str):
            unitstr_v = UnitStr.from_str(v)

            if len(unitstr_v.units_str) == 1 and isinstance(unitstr_v.units_str[0], str):
                new_kwargs[k] = unitstr_v.units_str[0]
            else:
                new_kwargs[k] = unitstr_v

        elif isinstance(v, BaseText):
            new_kwargs[k] = v

        else:
            raise ValueError(
                "Column labels must be strings or BaseText objects. Use `md()` or `html()` for formatting."
            )

    boxhead = self._boxhead._set_column_labels(new_kwargs)

    return self._replace(_boxhead=boxhead)


def cols_align(self: GTSelf, align: str = "left", columns: SelectExpr = None) -> GTSelf:
    """
    Set the alignment of one or more columns.

    The `cols_align()` method sets the alignment of one or more columns. The `align` argument
    can be set to one of `"left"`, `"center"`, or `"right"` and the `columns` argument can be
    used to specify which columns to apply the alignment to. If `columns` is not specified, the
    alignment is applied to all columns.

    Parameters
    ----------
    align
        The alignment to apply. Must be one of `"left"`, `"center"`, or `"right"`.
    columns
        The columns to target. Can either be a single column name or a series of column names
        provided in a list. If `None`, the alignment is applied to all columns.

    Returns
    -------
    GT
        The GT object is returned. This is the same object that the method is called on so that we
        can facilitate method chaining.

    Examples
    --------
    Let's use the `countrypops` to create a small table. We can change the alignment of the
    `population` column with `cols_align()`. In this example, the column label and body cells of
    `population` will be aligned to the left.

    ```{python}
    from great_tables import GT
    from great_tables.data import countrypops

    countrypops_mini = countrypops.loc[countrypops["country_name"] == "San Marino"][
        ["country_name", "year", "population"]
    ].tail(5)

    (
        GT(countrypops_mini, rowname_col="year", groupname_col="country_name")
        .cols_align(align="left", columns="population")
    )
    ```

    """

    # Throw if `align` is not one of the three allowed values
    if align not in ["left", "center", "right"]:
        raise ValueError("Align must be one of 'left', 'center', or 'right'.")

    # Get the full list of column names for the data
    column_names = self._boxhead._get_columns()

    # Upgrade `columns` to a list if `columns` is a string and not None
    if isinstance(columns, str):
        columns = [columns]
        _assert_list_is_subset(columns, set_list=column_names)
    elif columns is None:
        columns = column_names

    sel_cols = resolve_cols_c(data=self, expr=columns)

    # Set the alignment for each column
    return self._replace(_boxhead=self._boxhead._set_column_aligns(sel_cols, align=align))
