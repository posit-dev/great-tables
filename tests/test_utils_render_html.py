import pandas as pd
import polars as pl
from great_tables import GT, exibble, html, loc, md, style
from great_tables._utils_render_html import (
    create_body_component_h,
    create_columns_component_h,
    create_heading_component_h,
    create_source_notes_component_h,
)

small_exibble = exibble[["num", "char"]].head(3)


def assert_rendered_source_notes(snapshot, gt):
    built = gt._build_data("html")
    source_notes = create_source_notes_component_h(built)

    assert snapshot == source_notes


def assert_rendered_heading(snapshot, gt):
    built = gt._build_data("html")
    heading = create_heading_component_h(built)

    assert snapshot == heading


def assert_rendered_columns(snapshot, gt):
    built = gt._build_data("html")
    columns = create_columns_component_h(built)

    assert snapshot == str(columns)


def assert_rendered_body(snapshot, gt):
    built = gt._build_data("html")
    body = create_body_component_h(built)

    assert snapshot == body


def test_source_notes_snap(snapshot):
    new_gt = (
        GT(exibble)
        .tab_source_note(md("An **important** note."))
        .tab_source_note(md("Another *important* note."))
        .tab_source_note("A plain note.")
        .tab_source_note(html("An <strong>HTML heavy</strong> note."))
    )

    assert_rendered_source_notes(snapshot, new_gt)


def test_render_groups_reordered(snapshot):
    df = pd.DataFrame(
        {"row": [0, 1, 2, 3], "g": ["A", "B", "A", "B"], "x": ["00", "11", "22", "33"]}
    )

    new_gt = GT(df, rowname_col="row", groupname_col="g")

    assert_rendered_body(snapshot, new_gt)


def test_body_multiple_locations(snapshot):
    new_gt = GT(small_exibble).tab_style(
        style=style.fill(color="red"),
        locations=[
            loc.body(columns="num", rows=[0, 2]),
            loc.body(columns="char", rows=[1]),
        ],
    )

    assert_rendered_body(snapshot, new_gt)


def test_body_multiple_styles(snapshot):
    new_gt = GT(small_exibble).tab_style(
        style=[style.fill(color="red"), style.borders("left")],
        locations=loc.body(columns="num", rows=[0]),
    )

    assert_rendered_body(snapshot, new_gt)


def test_styling_data_01(snapshot):
    new_gt = GT(small_exibble).tab_style(
        style=style.text(color="red"),
        locations=loc.body(),
    )

    assert_rendered_body(snapshot, new_gt)


def test_styling_data_02(snapshot):
    new_gt = GT(small_exibble).tab_style(
        style=style.text(color="red"),
        locations=loc.body(columns=["char"]),
    )

    assert_rendered_body(snapshot, new_gt)


def test_styling_data_03(snapshot):
    new_gt = GT(small_exibble).tab_style(
        style=style.text(color="red"),
        locations=loc.body(columns="char", rows=[0, 2]),
    )

    assert_rendered_body(snapshot, new_gt)


def test_styling_data_04(snapshot):
    new_gt = GT(small_exibble).tab_style(
        style=style.text(color="red"),
        locations=loc.body(columns=[], rows=[0, 2]),
    )

    assert_rendered_body(snapshot, new_gt)


def test_styling_data_05(snapshot):
    new_gt = GT(small_exibble).tab_style(
        style=style.text(color="red"),
        locations=loc.body(columns="char", rows=[]),
    )

    assert_rendered_body(snapshot, new_gt)


def test_styling_data_06(snapshot):
    new_gt = GT(small_exibble).tab_style(
        style=style.text(color="red"),
        locations=loc.body(columns=[], rows=[]),
    )

    assert_rendered_body(snapshot, new_gt)


def test_styling_data_07(snapshot):
    new_gt = GT(small_exibble).tab_style(
        style=style.borders(sides="left"),
        locations=loc.body(columns="char", rows=[0, 2]),
    )

    assert_rendered_body(snapshot, new_gt)


def test_styling_data_08(snapshot):
    new_gt = GT(small_exibble).tab_style(
        style=style.borders(sides=["left"]),
        locations=loc.body(columns="char", rows=[0, 2]),
    )

    assert_rendered_body(snapshot, new_gt)


def test_styling_data_09(snapshot):
    new_gt = GT(small_exibble).tab_style(
        style=style.borders(sides=["left", "right"]),
        locations=loc.body(columns="char", rows=[0, 2]),
    )

    assert_rendered_body(snapshot, new_gt)


def test_styling_data_10(snapshot):
    new_gt = GT(small_exibble).tab_style(
        style=style.borders(sides="all"),
        locations=loc.body(columns="char", rows=[0, 2]),
    )

    assert_rendered_body(snapshot, new_gt)


def test_render_polars_list_col(snapshot):
    gt = GT(pl.DataFrame({"x": [[1, 2]]}))

    assert_rendered_body(snapshot, gt)


def test_multiple_spanners_pads_for_stubhead_label(snapshot):
    # NOTE: see test_spanners.test_multiple_spanners_above_one
    gt = (
        GT(exibble, rowname_col="row", groupname_col="group")
        .tab_spanner("A", ["num", "char", "fctr"])
        .tab_spanner("B", ["fctr"])
        .tab_spanner("C", ["num", "char"])
        .tab_spanner("D", ["fctr", "date", "time"])
        .tab_spanner("E", spanners=["B", "C"])
        .tab_stubhead(label="Group")
    )

    assert_rendered_columns(snapshot, gt)


# Location style rendering -------------------------------------------------------------------------
# these tests focus on location classes being correctly picked up
def test_loc_column_labels():
    gt = GT(pl.DataFrame({"x": [1], "y": [2]}))

    new_gt = gt.tab_style(style.fill("yellow"), loc.column_labels(columns=["x"]))
    el = create_columns_component_h(new_gt._build_data("html"))

    assert el.name == "tr"
    assert el.children[0].attrs["style"] == "background-color: yellow;"
    assert "style" not in el.children[1].attrs


def test_loc_kitchen_sink(snapshot):
    gt = (
        GT(exibble.loc[[0], ["num", "char", "fctr", "row", "group"]])
        .tab_header("title", "subtitle")
        .tab_stub(rowname_col="row", groupname_col="group")
        .tab_source_note("yo")
        .tab_spanner("spanner", ["char", "fctr"])
        .tab_stubhead("stubhead")
    )

    new_gt = (
        gt.tab_style(style.css("BODY"), loc.body())
        # Columns -----------
        .tab_style(style.css("COLUMN_LABEL"), loc.column_labels(columns="num"))
        .tab_style(style.css("COLUMN_HEADER"), loc.column_header())
        .tab_style(style.css("SPANNER_LABEL"), loc.spanner_labels(ids=["spanner"]))
        # Header -----------
        .tab_style(style.css("HEADER"), loc.header())
        .tab_style(style.css("SUBTITLE"), loc.subtitle())
        .tab_style(style.css("TITLE"), loc.title())
        # Footer -----------
        .tab_style(style.css("FOOTER"), loc.footer())
        .tab_style(style.css("SOURCE_NOTES"), loc.source_notes())
        # .tab_style(style.css("AAA"), loc.footnotes())
        # Stub --------------
        .tab_style(style.css("GROUP_LABEL"), loc.row_groups())
        .tab_style(style.css("STUB"), loc.stub())
        .tab_style(style.css("ROW_LABEL"), loc.stub(rows=[0]))
        .tab_style(style.css("STUBHEAD"), loc.stubhead())
    )

    html = new_gt.as_raw_html()
    cleaned = html[html.index("<table") :]
    assert cleaned == snapshot
