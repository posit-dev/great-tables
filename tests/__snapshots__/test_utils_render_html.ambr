# serializer version: 1
# name: test_body_multiple_locations
  '''
  <tbody class="gt_table_body">
    <tr>
      <td style="background-color: red;" class="gt_row gt_right">0.1111</td>
      <td class="gt_row gt_left">apricot</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">2.222</td>
      <td style="background-color: red;" class="gt_row gt_left">banana</td>
    </tr>
    <tr>
      <td style="background-color: red;" class="gt_row gt_right">33.33</td>
      <td class="gt_row gt_left">coconut</td>
    </tr>
  </tbody>
  '''
# ---
# name: test_body_multiple_styles
  '''
  <tbody class="gt_table_body">
    <tr>
      <td style="background-color: red; border-left: 1px solid #000000;" class="gt_row gt_right">0.1111</td>
      <td class="gt_row gt_left">apricot</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">2.222</td>
      <td class="gt_row gt_left">banana</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">33.33</td>
      <td class="gt_row gt_left">coconut</td>
    </tr>
  </tbody>
  '''
# ---
# name: test_loc_kitchen_sink
  '''
  <table class="gt_table" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
  <thead>
  
    <tr class="gt_heading">
      <td colspan="4" class="gt_heading gt_title gt_font_normal" style="HEADER TITLE">title</td>
    </tr>
    <tr class="gt_heading">
      <td colspan="4" class="gt_heading gt_subtitle gt_font_normal gt_bottom_border" style="HEADER SUBTITLE">subtitle</td>
    </tr>
  <tr class="gt_col_headings gt_spanner_row">
    <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="2" colspan="1" style="STUBHEAD" scope="col" id="stubhead">stubhead</th>
    <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="2" colspan="1" style="COLUMN_HEADER COLUMN_LABEL" scope="col" id="num">num</th>
    <th class="gt_center gt_columns_top_border gt_column_spanner_outer" rowspan="1" colspan="2" style="COLUMN_HEADER SPANNER_LABEL" scope="colgroup" id="spanner">
      <span class="gt_column_spanner">spanner</span>
    </th>
  </tr>
  <tr class="gt_col_headings">
    <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" style="COLUMN_HEADER" scope="col" id="char">char</th>
    <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" style="COLUMN_HEADER" scope="col" id="fctr">fctr</th>
  </tr>
  </thead>
  <tbody class="gt_table_body">
    <tr class="gt_group_heading_row">
      <th class="gt_group_heading" colspan="4" style="GROUP_LABEL">grp_a</th>
    </tr>
    <tr>
      <th style="STUB ROW_LABEL" class="gt_row gt_left gt_stub">row_1</th>
      <td style="BODY" class="gt_row gt_right">0.1111</td>
      <td style="BODY" class="gt_row gt_left">apricot</td>
      <td style="BODY" class="gt_row gt_left">one</td>
    </tr>
  </tbody>
    <tfoot class="gt_sourcenotes">
    
    <tr>
      <td class="gt_sourcenote" colspan="4" style="FOOTER SOURCE_NOTES">yo</td>
    </tr>
  
  </tfoot>
  
  </table>
  
  </div>
          
  '''
# ---
# name: test_multiple_spanners_pads_for_stubhead_label
  '''
  <tr class="gt_col_headings gt_spanner_row">
    <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col">
      <span>&nbsp</span>
    </th>
    <th class="gt_center gt_columns_bottom_border gt_columns_top_border gt_column_spanner_outer" rowspan="1" colspan="3" scope="colgroup">
      <span class="gt_column_spanner">E</span>
    </th>
    <th class="gt_center gt_columns_bottom_border gt_columns_top_border gt_column_spanner_outer" rowspan="1" colspan="4" scope="colgroup">
      <span>&nbsp;</span>
    </th>
  </tr>
  <tr class="gt_col_headings gt_spanner_row">
    <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col">
      <span>&nbsp</span>
    </th>
    <th class="gt_center gt_columns_bottom_border gt_columns_top_border gt_column_spanner_outer" rowspan="1" colspan="2" scope="colgroup">
      <span>&nbsp;</span>
    </th>
    <th class="gt_center gt_columns_bottom_border gt_columns_top_border gt_column_spanner_outer" rowspan="1" colspan="3" scope="colgroup">
      <span class="gt_column_spanner">D</span>
    </th>
    <th class="gt_center gt_columns_bottom_border gt_columns_top_border gt_column_spanner_outer" rowspan="1" colspan="2" scope="colgroup">
      <span>&nbsp;</span>
    </th>
  </tr>
  <tr class="gt_col_headings gt_spanner_row">
    <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col">
      <span>&nbsp</span>
    </th>
    <th class="gt_center gt_columns_bottom_border gt_columns_top_border gt_column_spanner_outer" rowspan="1" colspan="2" scope="colgroup">
      <span class="gt_column_spanner">C</span>
    </th>
    <th class="gt_center gt_columns_bottom_border gt_columns_top_border gt_column_spanner_outer" rowspan="1" colspan="1" scope="col">
      <span class="gt_column_spanner">B</span>
    </th>
    <th class="gt_center gt_columns_bottom_border gt_columns_top_border gt_column_spanner_outer" rowspan="1" colspan="4" scope="colgroup">
      <span>&nbsp;</span>
    </th>
  </tr>
  <tr class="gt_col_headings gt_spanner_row">
    <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="2" colspan="1" scope="col" id="Group">Group</th>
    <th class="gt_center gt_columns_top_border gt_column_spanner_outer" rowspan="1" colspan="3" scope="colgroup" id="A">
      <span class="gt_column_spanner">A</span>
    </th>
    <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="2" colspan="1" scope="col" id="date">date</th>
    <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="2" colspan="1" scope="col" id="time">time</th>
    <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="2" colspan="1" scope="col" id="datetime">datetime</th>
    <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="2" colspan="1" scope="col" id="currency">currency</th>
  </tr>
  <tr class="gt_col_headings">
    <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="num">num</th>
    <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="char">char</th>
    <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="fctr">fctr</th>
  </tr>
  '''
# ---
# name: test_render_groups_reordered
  '''
  <tbody class="gt_table_body">
    <tr class="gt_group_heading_row">
      <th class="gt_group_heading" colspan="2">A</th>
    </tr>
    <tr>
      <th class="gt_row gt_left gt_stub">0</th>
      <td class="gt_row gt_right">00</td>
    </tr>
    <tr>
      <th class="gt_row gt_left gt_stub">2</th>
      <td class="gt_row gt_right">22</td>
    </tr>
    <tr class="gt_group_heading_row">
      <th class="gt_group_heading" colspan="2">B</th>
    </tr>
    <tr>
      <th class="gt_row gt_left gt_stub">1</th>
      <td class="gt_row gt_right">11</td>
    </tr>
    <tr>
      <th class="gt_row gt_left gt_stub">3</th>
      <td class="gt_row gt_right">33</td>
    </tr>
  </tbody>
  '''
# ---
# name: test_render_polars_list_col
  '''
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_center">[1, 2]</td>
    </tr>
  </tbody>
  '''
# ---
# name: test_source_notes_snap
  '''
    <tfoot class="gt_sourcenotes">
    
    <tr>
      <td class="gt_sourcenote" colspan="9">An <strong>important</strong> note.</td>
    </tr>
  
  
    <tr>
      <td class="gt_sourcenote" colspan="9">Another <em>important</em> note.</td>
    </tr>
  
  
    <tr>
      <td class="gt_sourcenote" colspan="9">A plain note.</td>
    </tr>
  
  
    <tr>
      <td class="gt_sourcenote" colspan="9">An <strong>HTML heavy</strong> note.</td>
    </tr>
  
  </tfoot>
  '''
# ---
# name: test_styling_data_01
  '''
  <tbody class="gt_table_body">
    <tr>
      <td style="color: red;" class="gt_row gt_right">0.1111</td>
      <td style="color: red;" class="gt_row gt_left">apricot</td>
    </tr>
    <tr>
      <td style="color: red;" class="gt_row gt_right">2.222</td>
      <td style="color: red;" class="gt_row gt_left">banana</td>
    </tr>
    <tr>
      <td style="color: red;" class="gt_row gt_right">33.33</td>
      <td style="color: red;" class="gt_row gt_left">coconut</td>
    </tr>
  </tbody>
  '''
# ---
# name: test_styling_data_02
  '''
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_right">0.1111</td>
      <td style="color: red;" class="gt_row gt_left">apricot</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">2.222</td>
      <td style="color: red;" class="gt_row gt_left">banana</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">33.33</td>
      <td style="color: red;" class="gt_row gt_left">coconut</td>
    </tr>
  </tbody>
  '''
# ---
# name: test_styling_data_03
  '''
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_right">0.1111</td>
      <td style="color: red;" class="gt_row gt_left">apricot</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">2.222</td>
      <td class="gt_row gt_left">banana</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">33.33</td>
      <td style="color: red;" class="gt_row gt_left">coconut</td>
    </tr>
  </tbody>
  '''
# ---
# name: test_styling_data_04
  '''
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_right">0.1111</td>
      <td class="gt_row gt_left">apricot</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">2.222</td>
      <td class="gt_row gt_left">banana</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">33.33</td>
      <td class="gt_row gt_left">coconut</td>
    </tr>
  </tbody>
  '''
# ---
# name: test_styling_data_05
  '''
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_right">0.1111</td>
      <td class="gt_row gt_left">apricot</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">2.222</td>
      <td class="gt_row gt_left">banana</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">33.33</td>
      <td class="gt_row gt_left">coconut</td>
    </tr>
  </tbody>
  '''
# ---
# name: test_styling_data_06
  '''
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_right">0.1111</td>
      <td class="gt_row gt_left">apricot</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">2.222</td>
      <td class="gt_row gt_left">banana</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">33.33</td>
      <td class="gt_row gt_left">coconut</td>
    </tr>
  </tbody>
  '''
# ---
# name: test_styling_data_07
  '''
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_right">0.1111</td>
      <td style="border-left: 1px solid #000000;" class="gt_row gt_left">apricot</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">2.222</td>
      <td class="gt_row gt_left">banana</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">33.33</td>
      <td style="border-left: 1px solid #000000;" class="gt_row gt_left">coconut</td>
    </tr>
  </tbody>
  '''
# ---
# name: test_styling_data_08
  '''
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_right">0.1111</td>
      <td style="border-left: 1px solid #000000;" class="gt_row gt_left">apricot</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">2.222</td>
      <td class="gt_row gt_left">banana</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">33.33</td>
      <td style="border-left: 1px solid #000000;" class="gt_row gt_left">coconut</td>
    </tr>
  </tbody>
  '''
# ---
# name: test_styling_data_09
  '''
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_right">0.1111</td>
      <td style="border-left: 1px solid #000000;border-right: 1px solid #000000;" class="gt_row gt_left">apricot</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">2.222</td>
      <td class="gt_row gt_left">banana</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">33.33</td>
      <td style="border-left: 1px solid #000000;border-right: 1px solid #000000;" class="gt_row gt_left">coconut</td>
    </tr>
  </tbody>
  '''
# ---
# name: test_styling_data_10
  '''
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_right">0.1111</td>
      <td style="border-top: 1px solid #000000;border-bottom: 1px solid #000000;border-left: 1px solid #000000;border-right: 1px solid #000000;" class="gt_row gt_left">apricot</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">2.222</td>
      <td class="gt_row gt_left">banana</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">33.33</td>
      <td style="border-top: 1px solid #000000;border-bottom: 1px solid #000000;border-left: 1px solid #000000;border-right: 1px solid #000000;" class="gt_row gt_left">coconut</td>
    </tr>
  </tbody>
  '''
# ---
